_examples/hellocolibri/hellocolibri.go

```diff
@@ -0,0 +1,340 @@
+// Copyright 2021 ETH Zurich
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"crypto/rand"
+	"fmt"
+	"net"
+	"strings"
+	"time"
+
+	"github.com/scionproto/scion/go/lib/addr"
+	libcol "github.com/scionproto/scion/go/lib/colibri"
+	colapi "github.com/scionproto/scion/go/lib/colibri/client"
+	"github.com/scionproto/scion/go/lib/colibri/reservation"
+	"github.com/scionproto/scion/go/lib/common"
+	"github.com/scionproto/scion/go/lib/sciond"
+	colpath "github.com/scionproto/scion/go/lib/slayers/path/colibri"
+	"github.com/scionproto/scion/go/lib/snet"
+	"github.com/scionproto/scion/go/lib/sock/reliable"
+)
+
+const (
+	serverSciondPath = "[fd00:f00d:cafe::7f00:c]:30255" // AS 1-ff00:0:112
+	clientSciondPath = "127.0.0.20:30255"               // AS 1-ff00:0:111
+	serverPort       = 12345
+)
+
+func main() {
+	serverChan := make(chan []byte)
+	server(serverChan)
+	// client(serverChan)
+	fmt.Println("==============================================================================================")
+	// after client is finished, run the extended API example
+	clientExtendedAPI(serverChan)
+}
+
+func server(serverChan chan []byte) {
+	fmt.Println("server starting")
+	ctx, cancelF := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancelF()
+
+	daemon, err := sciond.NewService(serverSciondPath).Connect(ctx)
+	check(err)
+
+	localIA, err := daemon.LocalIA(ctx)
+	check(err)
+	udpAddr, err := net.ResolveUDPAddr("udp", serverSciondPath)
+	check(err)
+	udpAddr.Port = serverPort
+
+	// init network
+	dispatcher := reliable.NewDispatcher(reliable.DefaultDispPath)
+	scionNet := snet.NewNetwork(localIA, dispatcher, sciond.RevHandler{Connector: daemon})
+
+	fmt.Printf("server at %s\n", udpAddr)
+	conn, err := scionNet.Listen(context.Background(), "udp", udpAddr, addr.SvcNone)
```

Comment:
The listener just always auto acks reservations? Or is there a method to handle and respond to reservation requests?

Reply:
That's still a `TODO` in the code (and the design). The initial idea is this: 
- The listener tells `sciond` or directly `SvcCOL` it accepts connections. It can blacklist sources as well.
- This `listen` message stating it accepts reservations expires after e.g. 60 seconds, and has to be renewed before it expires.
- When a reservation setup/renewal arrives to `SvcCOL`, it consults the table of destination endhosts and checks for blacklists and `listen` messages. It fails/accepts the reservation appropriately.

What I would really like to have in the API is a callback for the listener, that runs every time a reservation is received for setup or renewal in the destination endhost. But I cannot find a way to do so elegantly :disappointed: :panda_face: We should have a meeting to talk about this, and update the PR afterwards.

Reply:
What you mentioned above, passing a callback handler into `Listen()` for Colibri sounds reasonable to me. Or alternatively having a member/function on the type returned from `Listen()` that configures the reservation request handler.

This callback handler would be registered with sciond/colibri service for the specific port that the listener is listening on. If the listener is closed, the callback is unregistered. If a reservation request comes in to sciond/colibri service for a specific port, the callback registered for that port is called and the result (accept or deny) is used to reply to the reservation request. If no callback is registered the request is automatically denied. This sounds like what you were thinking of. Can you point out the problems of this approach?